### PR TITLE
[ML] Replace CResourceMonitor::m_AllocationFailures to reduce the memory footprint

### LINKED ISSUE #2524: [ML] Replace CResourceMonitor::m_AllocationFailures to reduce the memory footprint
The map `CResourceMonitor::m_AllocationFailures` collects the times of all allocation failures due to the hard_limit state. Although we only require the total number of failures and the time of the last failure in code, we keep the time of all failures in the map.

Hence, we can easily reduce the memory footprint of the anomaly jobs by refactoring this member variable.

### PR DESCRIPTION
**Pull request by valeriy42**
Fixes #2524

Mark it as a non-issue since it is not customer-facing.

### CODE REVIEW

File: include/model/CResourceMonitor.h [hunk 1]
```diff
@@ -259,7 +259,7 @@ class MODEL_EXPORT CResourceMonitor {
     TMemoryUsageReporterFunc m_MemoryUsageReporter;
 
     //! Keep track of times of anomaly detector allocation failures
-    TTimeSizeMap m_AllocationFailures;
+    TSizeTimePr m_AllocationFailures;
```
Comment (droberts195) [inline include/model/CResourceMonitor.h:262]: The comment is now wrong.

But I think the code would be much clearer overall if there were two separate member variables, namely `m_AllocationFailureCount` and `m_LastAllocationFailureTime`. Readers should be able to guess what those mean without even referring to the comment.

**Review by dimkots — APPROVED** [review]
LGTM 🚀 (But let's wait for a more exp. c++ person)

File: include/model/CResourceMonitor.h (hunk 1 above)
Comment (valeriy42) [reply include/model/CResourceMonitor.h]: This makes sense. Thank you.

**Review by droberts195 — APPROVED** [review]
LGTM

Thanks for renaming the variables. It's much clearer now than when it was using the `map` originally as well as being clearer than the `pair`.

File: include/model/CResourceMonitor.h [hunk 2]
```diff
@@ -258,8 +257,11 @@ class MODEL_EXPORT CResourceMonitor {
     //! Callback function to fire when memory usage increases by 1%
     TMemoryUsageReporterFunc m_MemoryUsageReporter;
 
-    //! Keep track of times of anomaly detector allocation failures
-    TTimeSizeMap m_AllocationFailures;
+    //! Keep track of the number of allocation failures
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/model/CResourceMonitor.h]: comment updated

**Pull request merged**